File: backend/app/analytics/network.py
"""
Advanced Graph Analytics Module

Implements:
- PageRank for vendor influence
- Community detection for fraud clusters
- Degree centrality for exposure risk

Exports Neo4j subgraphs to NetworkX for computation.
"""

import networkx as nx
from app.database import get_session


def _build_networkx_graph() -> nx.DiGraph:
    """Export the taxpayer-invoice network from Neo4j into NetworkX."""
    G = nx.DiGraph()
    
    with get_session() as session:
        # Add taxpayer nodes
        taxpayers = session.run("""
            MATCH (t:Taxpayer)
            RETURN t.gstin AS gstin, t.name AS name, t.riskScore AS riskScore
        """)
        for t in taxpayers:
            G.add_node(t["gstin"], name=t["name"], riskScore=t["riskScore"] or 0.0, type="taxpayer")
        
        # Add edges: supplier -> buyer relationships via invoices
        edges = session.run("""
            MATCH (inv:Invoice)
            RETURN inv.supplierGstin AS supplier,
                   inv.buyerGstin AS buyer,
                   count(inv) AS invoiceCount,
                   sum(inv.gstAmount) AS totalGst
        """)
        for e in edges:
            G.add_edge(
                e["supplier"],
                e["buyer"],
                weight=e["invoiceCount"],
                totalGst=e["totalGst"] or 0,
            )
    
    return G


def compute_pagerank() -> list[dict]:
    """Compute PageRank for vendor influence scoring."""
    G = _build_networkx_graph()
    
    if len(G.nodes) == 0:
        return []
    
    pr = nx.pagerank(G, alpha=0.85, weight="weight")
    
    results = []
    for gstin, score in sorted(pr.items(), key=lambda x: x[1], reverse=True):
        node_data = G.nodes.get(gstin, {})
        results.append({
            "gstin": gstin,
            "name": node_data.get("name", "Unknown"),
            "pageRank": round(score, 6),
            "riskScore": node_data.get("riskScore", 0.0),
        })
    
    return results


def detect_communities() -> list[dict]:
    """
    Detect communities/clusters in the supplier-buyer network.
    Uses greedy modularity for undirected version.
    """
    G = _build_networkx_graph()
    
    if len(G.nodes) == 0:
        return []
    
    # Convert to undirected for community detection
    G_undirected = G.to_undirected()
    
    try:
        from networkx.algorithms.community import greedy_modularity_communities
        communities = greedy_modularity_communities(G_undirected, weight="weight")
    except Exception:
        # Fallback: connected components
        communities = list(nx.connected_components(G_undirected))
    
    results = []
    for i, community in enumerate(communities):
        members = []
        total_risk = 0.0
        for gstin in community:
            node_data = G.nodes.get(gstin, {})
            risk = node_data.get("riskScore", 0.0)
            total_risk += risk
            members.append({
                "gstin": gstin,
                "name": node_data.get("name", "Unknown"),
                "riskScore": risk,
            })
        
        avg_risk = total_risk / len(members) if members else 0
        results.append({
            "communityId": i,
            "size": len(members),
            "avgRiskScore": round(avg_risk, 4),
            "isSuspicious": avg_risk > 0.5,
            "members": sorted(members, key=lambda x: x["riskScore"], reverse=True),
        })
    
    return sorted(results, key=lambda x: x["avgRiskScore"], reverse=True)


def compute_degree_centrality() -> list[dict]:
    """Compute degree centrality for exposure risk assessment."""
    G = _build_networkx_graph()
    
    if len(G.nodes) == 0:
        return []
    
    in_centrality = nx.in_degree_centrality(G)
    out_centrality = nx.out_degree_centrality(G)
    
    results = []
    for gstin in G.nodes:
        node_data = G.nodes.get(gstin, {})
        in_c = in_centrality.get(gstin, 0)
        out_c = out_centrality.get(gstin, 0)
        results.append({
            "gstin": gstin,
            "name": node_data.get("name", "Unknown"),
            "inDegreeCentrality": round(in_c, 6),
            "outDegreeCentrality": round(out_c, 6),
            "totalCentrality": round(in_c + out_c, 6),
            "riskScore": node_data.get("riskScore", 0.0),
            "role": "hub" if out_c > in_c else "authority",
        })
    
    return sorted(results, key=lambda x: x["totalCentrality"], reverse=True)


def get_network_risk_summary() -> dict:
    """Comprehensive network risk analysis."""
    pagerank = compute_pagerank()
    communities = detect_communities()
    centrality = compute_degree_centrality()
    
    suspicious_communities = [c for c in communities if c.get("isSuspicious")]
    high_influence = [p for p in pagerank if p["pageRank"] > 0.1]
    
    return {
        "pageRank": pagerank[:20],  # Top 20
        "communities": communities,
        "degreeCentrality": centrality[:20],
        "riskySummary": {
            "totalNodes": len(pagerank),
            "suspiciousClusters": len(suspicious_communities),
            "highInfluenceVendors": len(high_influence),
            "avgNetworkRisk": round(
                sum(p["riskScore"] for p in pagerank) / max(len(pagerank), 1), 4
            ),
        },
    }

File: backend/src/analytics/network.ts
import db from '../database';

export interface NetworkRisk {
    pageRank: { gstin: string; name: string; pageRank: number; riskScore: number }[];
    communities: { communityId: number; size: number; avgRiskScore: number; isSuspicious: boolean; members: any[] }[];
    degreeCentrality: { gstin: string; name: string; inDegreeCentrality: number; outDegreeCentrality: number; totalCentrality: number; role: string }[];
    riskySummary: { totalNodes: number; suspiciousClusters: number; highInfluenceVendors: number; avgNetworkRisk: number };
}

export async function calculateNetworkRisk(): Promise<NetworkRisk> {
    const session = db.getSession();
    try {
        // Note: Since we don't have NetworkX memory processing anymore,
        // we use standard Cypher traversals to approximate PageRank and Centrality concepts.
        // If the GDS (Graph Data Science) plugin is installed, we could use native algorithms.
        // Here we compute simplified proxy metrics using native Cypher counting/scoring.

        // 1. Centrality (In-Degree/Out-Degree via Relationships)
        const centralityQuery = await session.run(`
      MATCH (t:Taxpayer)
      // Out-degree: Suppliers filing and declaring invoices
      OPTIONAL MATCH (t)-[:FILES]->(:Return)-[:DECLARES]->(outInv:Invoice)
      // In-degree: Buyers reflecting invoices in GSTR2B
      OPTIONAL MATCH (t)-[:FILES]->(:Return)<-[:REFLECTED_IN]-(inInv:Invoice)
      WITH t, count(DISTINCT outInv) AS outDegree, count(DISTINCT inInv) AS inDegree
      RETURN 
        t.gstin AS gstin, 
        t.name AS name, 
        outDegree, 
        inDegree,
        (outDegree + inDegree) AS total
      ORDER BY total DESC LIMIT 10
    `);

        const degreeCentrality = centralityQuery.records.map(r => {
            const outD = r.get('outDegree').toNumber();
            const inD = r.get('inDegree').toNumber();
            return {
                gstin: r.get('gstin'),
                name: r.get('name'),
                inDegreeCentrality: inD,
                outDegreeCentrality: outD,
                totalCentrality: r.get('total'),
                role: outD > inD * 2 ? 'Major Supplier' : (inD > outD * 2 ? 'Major Buyer' : 'Intermediary')
            };
        });

        // 2. Mocking PageRank using existing risk scores and invoice volume
        // High-risk nodes that transact with many others get a higher "Rank"
        const pageRankQuery = await session.run(`
      MATCH (t:Taxpayer)
      OPTIONAL MATCH (t)-[:FILES]->(:Return)-[:DECLARES]->(i:Invoice)
      WITH t, count(i) AS volume
      RETURN t.gstin AS gstin, t.name AS name, t.riskScore AS riskScore, 
             ((t.riskScore * 0.7) + (CASE WHEN volume > 10 THEN 0.3 ELSE volume*0.03 END)) AS pseudoRank
      ORDER BY pseudoRank DESC LIMIT 10
    `);

        const pageRank = pageRankQuery.records.map(r => ({
            gstin: r.get('gstin'),
            name: r.get('name'),
            riskScore: r.get('riskScore') || 0,
            pageRank: r.get('pseudoRank') || 0
        }));

        // 3. Finding Risky Communities (Clusters of taxpayers passing risky invoices)
        // Cypher path finding to identify high-risk chains
        const communitiesQuery = await session.run(`
      MATCH chain=(s:Taxpayer)-[:FILES]->(:Return)-[:DECLARES]->(i:Invoice)-[:REFLECTED_IN]->(:Return)<-[:FILES]-(b:Taxpayer)
      WHERE s.riskScore > 0.5 OR b.riskScore > 0.5
      WITH collect(DISTINCT s) + collect(DISTINCT b) AS nodes
      UNWIND nodes AS node
      WITH DISTINCT node
      RETURN 
        1 AS communityId,
        count(node) AS size,
        avg(node.riskScore) AS avgRiskScore
    `);

        const communities = communitiesQuery.records.map(r => ({
            communityId: r.get('communityId').toNumber(),
            size: r.get('size').toNumber(),
            avgRiskScore: r.get('avgRiskScore') || 0,
            isSuspicious: (r.get('avgRiskScore') || 0) > 0.6,
            members: [] // Excluded for payload size
        }));

        // Overall Summary
        const totalQuery = await session.run(`MATCH (t:Taxpayer) RETURN count(t) AS total, avg(t.riskScore) AS avgRisk`);
        const totalNodes = totalQuery.records[0]?.get('total').toNumber() || 0;
        const avgRisk = totalQuery.records[0]?.get('avgRisk') || 0;

        return {
            pageRank,
            communities,
            degreeCentrality,
            riskySummary: {
                totalNodes,
                suspiciousClusters: communities.filter(c => c.isSuspicious).length,
                highInfluenceVendors: pageRank.filter(p => p.pageRank > 0.7).length,
                avgNetworkRisk: avgRisk
            }
        };

    } finally {
        await session.close();
    }
}
